src/client/iof/src/ionss/config.c

```diff
@@ -0,0 +1,504 @@
+		return -1;
+	}
+	return parse_boolean(&option->bool_val,
+			     (char *)node->data.scalar.value,
+			     (int)node->data.scalar.length,
+			     (char*[]) { "disable", "auto" });
+}
+
+static int set_flag(struct parsed_option_s *option,
+		    yaml_document_t *document, yaml_node_t *node)
+{
+	if (node->type != YAML_SCALAR_NODE) {
+		IOF_LOG_ERROR("Invalid YAML node type");
+		return -1;
+	}
+	return parse_boolean(&option->bool_val,
+			     (char *)node->data.scalar.value,
+			     (int)node->data.scalar.length,
+			     (char*[]) { "false", "true" });
+}
+
+static int set_string(struct parsed_option_s *option,
+		      yaml_document_t *document, yaml_node_t *node)
+{
+	int ret = 0;
+
+	if (node->type != YAML_SCALAR_NODE) {
+		IOF_LOG_ERROR("Invalid YAML node type");
+		return -1;
+	}
+	D_STRNDUP(option->ptr_val,
+		  ((char *)node->data.scalar.value),
+		  node->data.scalar.length);
+	if (option->ptr_val == NULL)
+		ret = -1;
+	else
+		IOF_LOG_DEBUG("Setting option value: %.*s",
+			      (int)node->data.scalar.length,
+			      (char *)node->data.scalar.value);
+	return ret;
+}
+
+static int parse_node(yaml_document_t *document, yaml_node_t *node,
+		      struct parsed_option_s *options, int num_options)
+{
+	int ret;
+	yaml_node_t *key_node, *val_node;
+	yaml_node_pair_t *node_pair;
+	struct parsed_option_s *sel_option;
+
+	if (node->type != YAML_MAPPING_NODE) {
+		IOF_LOG_ERROR("Invalid YAML node type");
+		return -1;
+	}
+	for (node_pair = node->data.mapping.pairs.start;
+	     node_pair < node->data.mapping.pairs.top; node_pair++) {
+		key_node = yaml_document_get_node(document, node_pair->key);
+		sel_option = find_option(options, num_options,
+					 (char *)key_node->data.scalar.value,
+					 key_node->data.scalar.length);
+		if (!sel_option) {
+			IOF_LOG_WARNING("Unknown configuration option %.*s",
+					(int)key_node->data.scalar.length,
+					(char *)key_node->data.scalar.value);
+			return -1;
+		}
+		IOF_LOG_DEBUG("Processing configuration option: %s",
+			      sel_option->key);
+		val_node = yaml_document_get_node(document, node_pair->value);
+		ret = sel_option->setter(sel_option, document, val_node);
+		if (ret) {
+			if (val_node->type == YAML_SCALAR_NODE) {
+				IOF_LOG_WARNING("Unknown configuration value %s %.*s",
+						sel_option->key,
+						(int)val_node->data.scalar.length,
+						(char *)val_node->data.scalar.value);
+			} else {
+				IOF_LOG_WARNING("Unknown configuration value %s",
+						sel_option->key);
+			}
+			return ret;
+		}
+		sel_option->is_set = true;
+	}
+	return 0;
+}
+
+static int parse_projections(struct parsed_option_s *option,
+			     yaml_document_t *document, yaml_node_t *node)
+{
+	int i, j, ret;
+	yaml_node_t *next_node;
+	yaml_node_item_t *node_item;
+	struct projections_s *proj = option->ptr_val;
+
+	if (node->type != YAML_SEQUENCE_NODE) {
+		IOF_LOG_ERROR("Invalid YAML node type");
+		return -1;
+	}
+	proj->num_options = 0;
+#define X(...)	++proj->num_options;
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop

(style)  macros should not use a trailing semicolon